include/exadg/operators/interior_penalty_parameter.h

```diff
@@ -25,65 +25,120 @@
+      return cell->global_active_cell_index();
+    else
+      return cell->global_level_cell_index();
+  };
 
-  for(unsigned int i = 0; i < n_cells; ++i)
   {
-    for(unsigned int v = 0; v < matrix_free.n_active_entries_per_cell_batch(i); ++v)
-    {
-      typename DoFHandler<dim>::cell_iterator cell = matrix_free.get_cell_iterator(i, v, dof_index);
-      fe_values.reinit(cell);
+    FaceIntegrator<dim, 1, Number> face_integrator(matrix_free, true, dof_index);
 
-      // calculate cell volume
-      Number volume = 0;
-      for(unsigned int q = 0; q < quadrature.size(); ++q)
+    unsigned int const n_face_batches =
+      matrix_free.n_inner_face_batches() + matrix_free.n_boundary_face_batches();
+    for(unsigned int face = 0; face < n_face_batches; ++face)
+    {
+      face_integrator.reinit(face);
+      VectorizedArray<Number> face_area = 0.0;
+      for(unsigned int q = 0; q < face_integrator.n_q_points; ++q)
+        face_area += face_integrator.JxW(q);
+
+      bool const is_interior_face = face < matrix_free.n_inner_face_batches();
+
+      // For interior faces, the face area is weighted by a factor of 0.5
+      // according to the formula by "K. Hillewaert, Development of the
+      // Discontinuous Galerkin Method for High-Resolution, Large Scale CFD
+      // and Acoustics in Industrial Geometries, Ph.D. thesis, Univ. de
+      // Louvain, 2013."
+      if(is_interior_face)
+        face_area = face_area * 0.5;
+
+      // Write the result to the vector entry corresponding to the respective
+      // cell slot. For boundary faces, we write it to only one cell,
+      // otherwise to both adjacent cells of the face.
+      for(unsigned int v = 0; v < matrix_free.n_active_entries_per_face_batch(face); ++v)
       {
-        volume += fe_values.JxW(q);
-      }
+        surface_areas(get_cell_index(matrix_free.get_face_iterator(face, v, true).first)) +=
+          face_area[v];
 
-      // calculate surface area
-      Number surface_area = 0;
-      for(const unsigned int f : cell->face_indices())
-      {
-        fe_face_values.reinit(cell, f);
-        Number const factor = (cell->at_boundary(f) && !cell->has_periodic_neighbor(f)) ? 1. : 0.5;
-        for(unsigned int q = 0; q < face_quadrature.size(); ++q)
-        {
-          surface_area += fe_face_values.JxW(q) * factor;
-        }
+        if(is_interior_face)
+          surface_areas(get_cell_index(matrix_free.get_face_iterator(face, v, false).first)) +=
+            face_area[v];
       }
+    }
+  }
 
-      array_penalty_parameter[i][v] = surface_area / volume;
+  surface_areas.compress(VectorOperation::add);
+  surface_areas.update_ghost_values();
+
+  // As a second step, we use a CellInterator to compute the volume and then
+  // combine it with the surface areas computed before
+  unsigned int const n_cells = matrix_free.n_cell_batches() + matrix_free.n_ghost_cell_batches();
+  array_penalty_parameter.resize(n_cells);
+  {
```

Comment:
Is there a reason that you open two scopes in this function? It does not really make sense in my opinion that you define `n_face_batches` inside the scope but `n_cells` outside the scope. I think I would prefer to remove the scopes.

Reply:
I agree, the scopes are not necessary (I had when I originally had more temporary variables that would have clashed between cell and face loops, but now it's only the evaluator which is unambiguous). Done in 5c06970b